tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/KnowledgeGuardian.scala

```diff
@@ -0,0 +1,265 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js tools             **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013-2016, LAMP/EPFL   **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-js.org/       **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+
+
+package org.scalajs.core.tools.linker.backend.emitter
+
+import scala.collection.mutable
+
+import org.scalajs.core.ir.ClassKind
+import org.scalajs.core.ir.Trees.FieldDef
+
+import org.scalajs.core.tools.linker._
+
+private[emitter] object KnowledgeGuardian {
+  private[KnowledgeGuardian] trait Unregisterable {
+    def unregister(invalidatable: Invalidatable): Unit
+  }
+
+  trait Invalidatable {
+    private[KnowledgeGuardian] def registeredTo(
+        unregisterable: Unregisterable): Unit
+
+    def invalidate(): Unit
+  }
```

Comment:
Not necessarily, but then `Unregisterable.unregister` needs to take a `KnowledgeAccessor`. If `KnowledgeAccessor` goes inside the class `KnowledgeGuardian`, then `Unregisterable` must do too (which is fine), unless we type is as `KnowledgeGuardian#KnowledgeAccessor`.

Also, if `Invalidate` goes away, by the same argument we could get rid of `Unregisterable` since only `Class` implements it. But that has a weirder feel about it.

IMO keeping `Unregisterable` and `Invalidatable` separate is clearer.